### PR TITLE
Gutenberg: allow Jetpack users to enable SSO in the desktop app

### PR DESCRIPTION
**Pull request by nsakaimbo**
### Description

This PR includes changes to Calypso that allow Jetpack users to enable SSO and navigate to the Gutenberg editor in the desktop app. We use a newly-added `desktop` middleware in Calypso to transmit redux events to the desktop's renderer process. Please see ([PR](`https://github.com/Automattic/wp-desktop/pull/876`)) for how these events are used.

### Overview

In order to toggle SSO from within the desktop app:

1. Gutenberg controller: when authenticating, check if the user is in a desktop context and, if using a Jetpack site, whether SSO is enabled.
1. Within the desktop app, the user is presented with the option to proceed in an external browser, or enable SSO and continue in-app (if the user has site admin privileges)
1. If the user enables SSO, the desktop app toggles SSO with `activateModule`. In addition, Calypso's internal redux state for the site has to be refreshed with `requestSite` prior to proceeding (otherwise the `site` object will not reflect updated SSO status).

### Motivation/Context

It was [recently discovered](https://github.com/Automattic/wp-desktop/issues/640#issuecomment-621268932) that Jetpack users cannot use Gutenberg within the desktop app, unless SSO is enabled, and are otherwise left in an unusable state when trying to edit posts in the app.

### Alternatives Considered

- The initial take on solving this problem had the desktop app make educated guesses based on the redirect sent by the remote site, _after_ trying to navigate to the editor. This approach is messy (and brittle). Ideally this logic should live within Calypso as Calypso has all the necessary context to decide whether or not Gutenberg can be used.
- Within Calypso itself, the are numerous places from which an editor event can be triggered. Implementing this logic within the Gutenberg controller seemed like the most reasonable place for this logic, as all components that enable editing eventually route through this code.

### CODE REVIEW

File: client/gutenberg/editor/controller.js [hunk 1]
```diff
@@ -75,6 +75,18 @@ function waitForSiteIdAndSelectedEditor( context ) {
 	} );
 }
 
+// TODO: Make these state selectors
```
Comment (nsakaimbo) [inline client/gutenberg/editor/controller.js:78]: TODO: Need to move these two helpers to `state/sites/selectors`.

**Comment by pento** [comment]
Nice work, @nsakaimbo! I agree that Calypso is the best place to handle this, we can fix the UX on WordPress.com when Jetpack SSO isn't enabled, too.

So, here's how Jetpack auth works, with SSO enabled:
- User isn't logged in to their site, is redirected to wp-login.php.
- Jetpack sees that wp-login.php is loaded, and there's a flag being passed to tell it to automatically go through the login process.
- Jetpack sends a message directly to the WP.com API, to say that a login request is coming, then redirects to wordpress.com/login.
- As we're in the app, the user is already authed on wordpress.com.
- WordPress.com sends a message directly to the site, to say that authentication was successful, and redirects back to the site's wp-login.php.
- When wp-login.php is loaded this time, Jetpack confirms the message it got from WordPress.com, and sets the auth cookie, then redirects back to the editor.

We don't want to prevent these redirects, as they're required in both the browser and the app, to set the auth cookie on the Jetpack site.

If SSO is disabled:
- Jetpack site admins get three options: Cancel, Redirect to Site, Enable SSO and continue in app.
- All other Jetpack site users are automatically redirected to the site, since they have no ability to change the setting. 

The Enable SSO and Continue button should:
- Call `[activateModule( siteId, 'iso' )](https://github.com/Automattic/wp-calypso/blob/fceec0456dcf9e281990a53310a1771ef17ddcd0/client/state/jetpack/modules/actions.js#L24)`.
- When that succeeds, SSO is enabled, so load the editor iframe as normal. Don't catch the Jetpack redirects, they should do the auth properly now.

When we're in the app, "Redirect to Site" should be "open in browser". Perhaps the app could also show a message to contact their site admin to enable SSO.

**Comment by nsakaimbo** [comment]
@pento Have been trying to chip away at this, but so far am encountering a couple of issues:

- I'm assuming the `activateModule` request returns cookies of some sort? The desktop app is currently set up only to ingest/set cookies at application boot. I'm currently trying to understand how Jetpack authorizes WP.com accounts after the `activateModule` request is successful. Is a separate cookie returned? Or is this a change to an existing one?

- The `site` info (returned in the redux store) does not reflect updated SSO status. For example, enabling SSO in-app, even though successful, will still lead to the prompt when trying to navigate to the editor again (this is aside from the cookie/auth issue mentioned above). So somehow we need to trigger a refresh of the site info in the Redux store.

File: client/gutenberg/editor/controller.js [hunk 2]
```diff
@@ -94,13 +101,19 @@ export const authenticate = ( context, next ) => {
...
 
-	const isAuthenticated =
+	let isAuthenticated =
 		globalThis.sessionStorage.getItem( storageKey ) || // Previously authenticated.
 		! isSiteWpcomAtomic( state, siteId ) || // Simple sites users are always authenticated.
-		isEnabled( 'desktop' ) || // The desktop app can store third-party cookies.
+		isDesktop || // The desktop app can store third-party cookies.
 		context.query.authWpAdmin; // Redirect back from the WP Admin login page to Calypso.
+
+	if ( isDesktop && isJetpackSite( state, siteId ) && ! ssoEnabled( reduxGetState(), siteId ) ) {
```
Comment (nsakaimbo) [inline client/gutenberg/editor/controller.js:113]: Note: will remove the `reduxGetState` here and just use `state`. Was seeing if this made a difference to the "prompted twice" issue (it doesn't). 😕 

TL;DR - Combined with changes in the desktop, this logic works as expected, but for some reason `sso` status read from the store is stale after the first attempt, and only becomes correct after the second prompt to the user. I'm using the `requestSite` call to update the site's state - perhaps my understanding of what this method does is incorrect.

File: client/gutenberg/editor/controller.js [hunk 3]
```diff
@@ -17,13 +17,14 @@ import getGutenbergEditorUrl from 'state/selectors/get-gutenberg-editor-url';
...
 import { requestSelectedEditor } from 'state/selected-editor/actions';
-import { getSiteUrl, getSiteOption } from 'state/sites/selectors';
+import { getSiteUrl, getSiteOption, getSite, isJetpackSite } from 'state/sites/selectors';
 import isSiteWpcomAtomic from 'state/selectors/is-site-wpcom-atomic';
 import { isEnabled } from 'config';
 import { Placeholder } from './placeholder';
 import { makeLayout, render } from 'controller';
 import isSiteUsingCoreSiteEditor from 'state/selectors/is-site-using-core-site-editor';
 import getSiteEditorUrl from 'state/selectors/get-site-editor-url';
+import { reduxGetState } from 'lib/redux-bridge';
```
Comment (belcherj) [inline client/gutenberg/editor/controller.js:27]: Remove

File: client/gutenberg/editor/controller.js [hunk 4]
```diff
@@ -75,6 +76,12 @@ function waitForSiteIdAndSelectedEditor( context ) {
 	} );
 }
 
+function ssoEnabled( state, siteId ) {
+	const site = getSite( state, siteId );
+	const siteActiveModules = site.options.active_modules;
```
Comment (belcherj) [inline client/gutenberg/editor/controller.js:81]: site?.options?.active_modules;

File: client/gutenberg/editor/controller.js (hunk 2 above)
Comment (belcherj) [reply client/gutenberg/editor/controller.js:113]: Going to remove `reduxGetState()` and replace with `state`

File: client/gutenberg/editor/controller.js [hunk 5]
```diff
@@ -125,7 +138,20 @@ export const authenticate = ( context, next ) => {
 
 	const siteUrl = getSiteUrl( state, siteId );
 	const wpAdminLoginUrl = addQueryArgs( { redirect_to: returnUrl }, `${ siteUrl }/wp-login.php` );
-	window.location.replace( wpAdminLoginUrl );
+
+	if ( isDesktop ) {
+		window.dispatchEvent(
```
Comment (belcherj) [inline client/gutenberg/editor/controller.js:143]: Add redux event and push the event to desktop/lib/middleware

File: client/lib/desktop/index.js [hunk 6]
```diff
@@ -40,12 +43,31 @@ const Desktop = {
 		ipc.on( 'signout', this.onSignout.bind( this ) );
 		ipc.on( 'toggle-notification-bar', this.onToggleNotifications.bind( this ) );
 		ipc.on( 'page-help', this.onShowHelp.bind( this ) );
+		ipc.on( 'enable-site-option', this.desktopCommandEnableSiteOption.bind( this ) );
```
Comment (belcherj) [inline client/lib/desktop/index.js:46]: Is it possible to only add listeners when they are needed and remove them afterwards?

File: client/state/index.js [hunk 7]
```diff
@@ -57,6 +59,7 @@ export function createReduxStore( initialState, reducer = initialReducer ) {
 		isBrowser && require( './analytics/middleware.js' ).analyticsMiddleware,
 		isBrowser && require( './lib/middleware.js' ).default,
 		isAudioSupported && require( './audio/middleware.js' ).default,
+		isDesktop && require( '../../client/lib/desktop/middleware.js' ).default,
```
Comment (belcherj) [inline client/state/index.js:62]: Should this belong in the state folder at ./desktop/middleware.js

Comment (nsakaimbo) [reply client/state/index.js]: Yep - that sounds more consistent. Moved.

**Comment by nsakaimbo** [comment]
**Update**: Things are working! 🎉 Thanks for the guidance, @pento ! Still needs a minor bit of refactoring, but I think this is fairly close to final.

**Comment by nsakaimbo** [comment]
Update: ready for review.

**Comment by nsakaimbo** [comment]
I see an e2e failure in ci/wp-e2e-tests-full-mobile, but this error appears unrelated to this change:

`Sign up for a free WordPress.com site from the Jetpack new site page, and log in via a magic link @signup @email @env:mocha|executor:local`

File: client/lib/desktop/index.js (hunk 6 above)
Comment (nsakaimbo) [reply client/lib/desktop/index.js]: Yep - refactored accordingly.
